### DISCUSSION TITLE
How do i disable semantic tokens?

### DISCUSSION THREAD

**Discussion by mstendorf** [mrcjkb/rustaceanvim#135] [Q&A] [answered]
After rustaceanvim finish loading and the lsp is enabled it also changes the tokens for my color theme. Rust-tools did not do this.

I think I just need to disable the semantic tokens added by rustaceanvim, but how do I go about this?

**Comment by mrcjkb** [comment, marked as answer]
Hey :wave: 

There are two ways you can do this:

- If your colorscheme only has issues with Rust, I would suggest you look at the [rust-analyzer configuration options](https://rust-analyzer.github.io/manual.html#configurationhttps://rust-analyzer.github.io/manual.html#configuration).
  There are a bunch of `rust-analyzer.semanticHighlighting.*` options that you can disable individually.
- If you are facing this issue with multiple languages, you can disable the client capability:

```lua
vim.api.nvim_create_autocmd('LspAttach', {
  group = vim.api.nvim_create_augroup('UserLspConfig', {}),
  callback = function(ev)
    local client = vim.lsp.get_client_by_id(ev.data.client_id)
    client.server_capabilities.semanticTokensProvider = nil
  end,
})
```

There is an open issue to create default links for non-standard semantic tokens provided by rust-analyzer: https://github.com/mrcjkb/rustaceanvim/issues/134

**Reply by mstendorf** [reply]
Hi,

Thank you for the fast an helpful reply. The issue is only with rust and only with rustaceanvim, if I change back to rust-tools the issue is gone. This leads me to believe it's an interaction with rustaceanvim that needs to be disabled?

I am a little out of my depth here and I could be totally wrong :)

**Reply by mstendorf** [reply]
I figured out how to do it just for rust, so for anyone comming here this seems to work:
```lua
vim.api.nvim_create_autocmd({ "LspAttach" }, {
  desc = "Disable semantic tokens for rust",
  group = vim.api.nvim_create_augroup("disable-rustacean-colors", { clear = true }),
  callback = function(opts)
    if vim.bo[opts.buf].filetype == "rust" then
      local client = vim.lsp.get_client_by_id(opts.data.client_id)
      client.server_capabilities.semanticTokensProvider = nil
    end
  end,
})

```

**Reply by mrcjkb** [reply]
:thinking: If it wasn't an issue with rust-tools, I think I know what is causing it.
Should be fixed with https://github.com/mrcjkb/rustaceanvim/commit/73854d5e720d4ea135e7508d9d24fb82b4ec8223, and the 3.12.2 release.

**Reply by mstendorf** [reply]
That is the version I am on now and the issues persists :)

**Reply by mrcjkb** [reply]
Okay, them I'm out of ideas for now.

That was the only change compared to rust-tools with regard to semantic tokens.
AFICT, neither this plugin, nor rust-tools, nor nvim-lspconfig do anything to explicitly set any client capabilities or server configs that relate to semantic highlighting.

Are you sure you were using the same rust-analyzer version with both plugins?

**Reply by mstendorf** [reply]
I am pretty sure I am using the same version unless one of the plugins installs a rust-analyzer outside the cargo directory.
How do I validate what rust-analyzer the plugins call?

**Reply by mrcjkb** [reply]
`:checkhealth rustaceanvim` will output the version that this plugin sees.

I'm not sure how you would see it with rust-tools. Maybe with `:LspInfo`, since it goes through lspconfig?

**Reply by mstendorf** [reply]
Any way to see the path rustaceanvim read rust-analyzer from?

**Reply by mrcjkb** [reply]
`:echo $PATH`

**Reply by mstendorf** [reply]
Sorry for the confusion, am on an m1 macbook so the rustup version is not working and had both a homebrew and mason version installed. Cleaned it up and verified the same version for both plugins. Issue still persists only for rustaceanvim.

I have a solution to overwrite it for now, so I am happy. I will gladly assist in debugging further if you want to figure this out.
The theme with the issue is Tokyonight, easiest way to see it is returning an Ok in orange color until rustaceanvim finishes loading then it  turns the same blueish as everything else :)

Thanks a lot for all the answers and the help :)

**Reply by mrcjkb** [reply]
Thanks. I'll see if I can reproduce it 😃

**Reply by mrcjkb** [reply]
@mstendorf do the changed tokens not match the theme colors?

Semantic tokens changing the colours is the intended behaviour. What I'm trying to figure out is if there are any rust-specific (non-standard) highlight tokens that Neovim colorthemes can't handle, or if this is an issue with other language servers that support semantic tokens.

In the latter case, you'll probably want to disable semantic highlighting anyway.

**Reply by mstendorf** [reply]
This is how it looks with rust-tools or when i disable semantic tokens in rustaceanvim.
![image](https://github.com/mrcjkb/rustaceanvim/assets/9951370/8bccc8f6-09a1-4d5a-b62a-8db41f9de3fb)

This is how it looks in rustaceanvim with semantic tokens enabled.
![image](https://github.com/mrcjkb/rustaceanvim/assets/9951370/43930472-2d80-4c91-bc90-ea655b9a4985)

I don't know if this anwers your question?

**Reply by mrcjkb** [reply]
Yes, it almost does :smile: :pray:

If you have time (whenever, not urgent), could you please hover over the `args`  and `customer_args` declarations, and run `:Inspect`, so I can see which semantic tokens they are getting?

Which variant of tokyonight are you using?

**Comment by mstendorf** [comment]
args:
![image](https://github.com/mrcjkb/rustaceanvim/assets/9951370/783a818c-fa9a-4a6b-98e5-e928e5af8ebb)

customer_args:
![image](https://github.com/mrcjkb/rustaceanvim/assets/9951370/01f91d31-e96f-4ab4-be72-7467a150f105)


I am using the tokyonight theme that comes stock with NVchad.

**Comment by dustypomerleau** [comment]
There are definitely some highlight groups that are specific to Rust Analyzer—like `@lsp.type.selfKeyword`, `@lsp.type.selfTypeKeyword`, `@lsp.type.generic`, and `@lsp.type.lifetime`—but the same issue exists in Treesitter scopes. At the end of the day, targeting those scopes is the theme designer's responsibility, and shouldn't fall in @mrcjkb's lap, particularly since they are a moving target and periodically change.

I hate to see you giving up on semantic highlighting—which easily distinguishes enums, structs, and traits, for example—just because a few highlight groups are not present in your theme. The number of `@lsp...` groups you need to override is relatively small. You can use [my theme](https://github.com/dustypomerleau/tol.nvim/blob/d49ac168c14c4ae5984cbed086bc81cd28b6a3f9/lua/tol/theme.lua#L406) for reference.
